File: consecutive-substrings-master/javascript/consecutive_substrings.js
function consecutiveSubstrings(string) {
  const substrings = [];

  for (let i = 0; i < string.length; i++) {
    for (let j = i + 1; j <= string.length; j++) {
      substrings.push(string.slice(i, j));
    }
  }

  return substrings;
}

// Test cases
console.log(consecutiveSubstrings('abc')); // ['a', 'ab', 'abc', 'b', 'bc', 'c']
console.log(consecutiveSubstrings('a')); // ['a']

if (require.main === module) {
  // add your own tests in here
  console.log("Expecting: ['a', 'ab', 'abc', 'b', 'bc', 'c']");
  console.log("=>", consecutiveSubstrings('abc'));

  console.log("");

  console.log("Expecting: ['a']");
  console.log("=>", consecutiveSubstrings('a'));
}

module.exports = consecutiveSubstrings;

// Please add your pseudocode to this file
// And a written explanation of your solution


if (require.main === module) {
  // add your own tests in here
  console.log("Expecting: ['a', 'ab', 'abc', 'b', 'bc', 'c']");
  console.log("=>", consecutiveSubstrings('abc'));

  console.log("");

  console.log("Expecting: ['a']");
  console.log("=>", consecutiveSubstrings('a'));
}

module.exports = consecutiveSubstrings;

// Please add your pseudocode to this file
// And a written explanation of your solution

// Create an empty array called substrings to store the consecutive substrings.
// Iterate over each character in the input string.
// Within the outer loop, iterate over each character starting from the current character to the end of the string.
// Append the substring from the current character to the end of the string to the substrings array.// Repeat steps 3-4 for all characters in the input string.
// Return the substrings array.

Explanation:
// We create an empty array called substrings to store the consecutive substrings.
// We iterate over each character in the input string using a for loop with the index variable i.
// Within the outer loop, we iterate over each character starting from the current character to the end of the string using another for loop with the index variable j.
// We use the slice() method to extract the substring from the current character i to the character at index j (exclusive) and push it into the substrings array.
// We repeat steps 3-4 for all characters in the input string.
// Finally, we return the substrings array.

File: solutions/consecutive_substrings.rb
def consecutive_substrings(str)
  substrings = []


  (0...str.length).each do |i|
    (i + 1..str.length).each do |j|
      substrings << str[i..j]
      
    end
  end

   substrings
end 

#Test caases
p consecutive_substrings('abc') # ['a', 'ab', 'abc', 'bc', 'c']
p consecutive_substrings('a') #{ ['a']


  string.split('').each_with_index do |char, idx|
    subs << char
    fragment = char

    string[idx + 1..-1].split('').each_with_index do |letter|
      fragment += letter
      subs << fragment
    end
  end

  subs




if __FILE__ == $PROGRAM_NAME
  puts "Expecting: ['a', 'ab', 'abc', 'b', 'bc', 'c']"
  print consecutive_substrings('abc')

  puts

  puts "Expecting: ['a']"
  print consecutive_substrings('a')

  # Don't forget to add your own!

  puts

  puts "Expecting: []"
  print consecutive_substrings('')

  puts

  puts "Expecting: ['a', 'ab', 'b']"
  print consecutive_substrings('ab')
end

# Please add your pseudocode to this file
#####################################################################################
 # initialize an empty array called subs
 # 
 # iterate over string:
 #    push character onto subs
 #    initialize fragment and store character in it
 # 
 #    iterate over string starting one index ahead:
 #      add letter to fragment
 #      push fragment onto subs
 # 
 # return subs
#####################################################################################

# And a written explanation of your solution
#####################################################################################
 # We can get all of the consecutive substrings by iterating over the string. We
 # push that character onto an array. We also store that character in a variable,
 # which we'll use to store the rest of the following characters. Next, we iterate
 # over the rest of the string. Each time we encounter a new character, we add it
 # onto the variable and then push the variable onto the array.
#####################################################################################
